README.md
## Argp++:

A C++ class library wrapper around the GNU C-language **argp** options 
parsing utility for command-line programs. The GNU documentation for 
the native _argp_ library is available here:

https://www.gnu.org/software/libc/manual/html_node/Argp.html

### Introduction:

For an excellent introduction to using GNU _argp_ in programs written
in C, please read:
    
**"Step-by-Step into Argp"**
(Copyright © 2010, 2015 by Ben Asselstine) 

which you can find here: http://www.nongnu.org/argpbook/

The **source code** in this package consists of four C++ files:
* include/argpp.hpp
* include/argppbaseimpl.hpp
* src/argpp.cpp
* src/argppbaseimpl.cpp

There is a dependency on the header `argp.h`, otherwise no additional
libraries are necessary to use the code.
    
### Installation:

For installation instructions for this package, please refer to the 
**INSTALL** file.
    
It is probably easiest just to add all the sources to a new or
existing project; then there is nothing to install. ;)
    
There are several demo programs using **argp++** located in the
**examples** subfolder. These are not built by default.
`make examples` will build them and leave the executables directly
in the `examples` folder if you have run `configure` and `make`.
You do NOT have to install `libargp++` as a shared library first in order
to build the example programs.

### Using Argp++ in your own programs:

Either way, using the sources directly or as a library, just add 
`#include "argpp.hpp"` (or `#include <argpp.hpp>` if you installed it 
that way) to your code **AFTER** any other `#include` lines (see below 
about bug...), define a class of your own derived from `argpp::ArgppBase`, 
implement at least the pure virtual function `ArgppBase::parserImpl` 
and any others you might need, and off you go! The demo programs in the
`examples` subfolder will get you started.
    
There are some traps to avoid when mixing C++ object variables and C code 
such as native _argp_ functions, which can call `exit` at times with total 
unawareness of any automatic C++ variable objects whose destructors might 
need to be called beforehand. Special care was taken not to fall into that
trap in designing the base class.
    
There is also a very annoying bug which was discovered when the `argp.h`
header file is included **before** some of the STL headers and compiling
with C++11 functionality. You can read about this here:
    
https://sourceware.org/bugzilla/show_bug.cgi?id=16907
    
Apparently, this has recently been fixed, but there are plenty of
platforms which might not be completely up-to-date here. Writing
`#include <argp.h>` as the last header included by `argpp.hpp`
seems to mitigate the effects of this bug.

NOTE: There is no **man** page and no **info** for this package. All 
documentation is contained either in the above-mentioned files, in the
`doc` folder, or in the source code itself. For more details, please 
consult the INSTALL file as well as the HTML documentation which was 
generated by **Doxygen**, available here:

http://www.stack.nl/~dimitri/doxygen/

### C++ language standard:

The header `argpp.hpp` has macros in place which will discover how 
the code is being compiled. If **C++11** support is available, it will 
be used by default. In particular, `std::unique_ptr<T>` is used to hold
a pointer to the root parser instance in a global variable. If C++11 is
**not** available, a substitute `DeleteHelper` class is used instead. 
Otherwise, none of the more advanced C++11 syntax is used in order to
make it possible to compile the code on older platforms (but hopefully
not TOO old...)
    
Please read the other information in the HTML docs (located in the 
`doc/html` subfolder) for further details about the design
and use of the library. The file "ChangeLog" will contain more details 
about the latest developments.

### Credits and license:

Project URL: https://github.com/bobhairgrove/Argp-plus-plus
    
Version 0.9.0 is the initial public release. Please see the files
COPYING and COPYING.LESSER for complete license details.

Written by Robert Hairgrove. Copyright (c) 2017 by Robert Hairgrove and
the Free Software Foundation. All feature requests and bug reports are
welcome and should be sent to Robert Hairgrove, <code@roberthairgrove.com>,
or you can create an issue at Github under the project URL mentioned above.

--

    This library is free software; you can redistribute it and/or modify it 
    under the terms of the GNU Lesser General Public License as published by 
    the Free Software Foundation; either version 2.1 of the License, or 
    (at your option) any later version. This library is distributed in the 
    hope that it will be useful, but WITHOUT ANY WARRANTY; without even the 
    implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
    See the GNU Lesser General Public License for more details. You should 
    have received a copy of the GNU Lesser General Public License along with 
    this library; if not, write to the Free Software Foundation, Inc., 
    51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA



NEWS
### Argp++ (version 0.9.0)

   This is the initial public release.

   A C++ wrapper class library around the GNU C-language **argp** options 
   parsing library for command-line programs.

   Project URL on Github:
   https://github.com/bobhairgrove/Argp-plus-plus
   
   Please read the files README, INSTALL, and the HTML docs in the 
   `doc` subfolder for further details about installation and use of the 
   library. The file `ChangeLog` will contain more details about the latest
   developments.
   
   Written by Robert Hairgrove. Copyright (c) 2017 by Robert Hairgrove and
   the Free Software Foundation. All feature requests and bug reports are
   welcome and should be sent to Robert Hairgrove <code@roberthairgrove.com>
   or create an issue at Github under the project URL mentioned above.

--

This library is free software; you can redistribute it and/or modify it 
under the terms of the GNU Lesser General Public License as published by 
the Free Software Foundation; either version 2.1 of the License, or 
(at your option) any later version. This library is distributed in the 
hope that it will be useful, but WITHOUT ANY WARRANTY; without even the 
implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU Lesser General Public License for more details. You should 
have received a copy of the GNU Lesser General Public License along with 
this library; if not, write to the Free Software Foundation, Inc., 
51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA


examples/ex_step_5/ex_step_5.cpp
/**
 * @file ex_step_5.cpp
 * @example ex_step_5.cpp
 * @brief An Argp-enabled program with an aliased long option.
 * @details Based on the book "Step-by-Step into Argp"
 * by Ben Asselstine (http://www.nongnu.org/argpbook/).
 * Takes each of the "step" examples and transforms them
 * into a C++ version using libargp++.
 */

/* ARGP++ is a C++ wrapper library around the GNU argp library.
   Copyright (C) 2017 by Robert Hairgrove <code@roberthairgrove.com>.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

   The argp library is part of the GNU C Library.
   Written by Miles Bader <miles@gnu.ai.mit.edu>.
   Copyright (C) 1995-1999, 2003, 2004, 2005, 2006, 2007, 2009
   Free Software Foundation, Inc.
*/

#include <iostream>
#include "../../include/argpp.hpp"

using namespace std;
using namespace argpp;

string prg_name;
static const string prg_version = "version 1.0";
static const string bug_addr    = "<code@roberthairgrove.com>";

extern void do_test();

//----------------------------------------------
/**
 * @brief The Argex_5 class
 */
class Argex_5 : public argpp::ArgppBase {
  public:
    Argex_5(int cargs, char** vargs, ArgppBase* parent)
      : ArgppBase(cargs, vargs, parent)
    {}
  private:
    error_t parserImpl(int key, const char *arg, argp_state *) {
      if (key == 'd') {
        unsigned int i;
        unsigned int dots = 0;
        if (arg == NULL) {
          dots = 1;
        } else {
          dots = strtoul(arg, NULL, 10);
        }
        for (i = 0; i < dots; i++) {
          cout << ".";
        }
        cout << endl;
        //------------------------------------
        // We could have recorded the parsed
        // option here, but we cannot distinguish
        // between 'd', "dot" and "period", so
        // we'll just let it ride...
        //------------------------------------
        return ARGPP_SUCCESS;
      }
      return ARGPP_UNKNOWN;
    }
};

//----------------------------------------------
bool setup_parsers(int cargc, char** vargv) {

  bool retval = false;
  ArgppOption opt;
  //-----------------------------------------------
  // We will add all of the options at once
  // by passing this vector:
  // the base class takes care of delimiting
  // the array of argp_option structs by appending
  // an empty argp_option when the time comes to
  // call the native argp_parse() function:
  //-----------------------------------------------
  ArgppOptions opts;

  opt.key_           = 'd';
  opt.long_name_     = "dot";
  opt.arg_           = "NUM";
  opt.flags_         = opt_arg_is_optional;
  opt.doc_or_header_ = "Show some dots on the screen";

  opts.push_back(opt);

  // 2nd option:
  opt.key_           = 0;
  opt.long_name_     = "period";
  opt.arg_           = "FOO";
  opt.flags_         = opt_is_alias;
  opt.doc_or_header_ = "Bar";

  opts.push_back(opt);


  // create the root instance:
  Argex_5* root = Factory<Argex_5>::createParser(cargc, vargv);

  if (root) {
    root->installProgVersion(prg_version);
    root->installBugAddress(bug_addr);

    // add options:
    retval = root->addOptions(opts);
  }
  return retval;
}

//----------------------------------------------
int main(int argc, char** argv) {
  prg_name = argv[0];
  if (setup_parsers(argc, argv)) {
    //----------------------------------------------
    // This function might not return ... see comments
    // in the file "boilerplate.cpp" and in the
    // documentation for argp++ for an explanation:
    //----------------------------------------------
    do_test();
  }
  return 0;
}
